lightning/src/ln/peers/handshake/acts.rs

```diff
@@ -1,42 +1,171 @@
+//! Act objects are thin wrappers about raw arrays for stack-based processing as well as convenient
+//! coercion to slices for flexible use by the higher-level modules.
 
-/// Wrapper for the second act message
-pub struct ActTwo(
-	pub(super) [u8; ACT_TWO_LENGTH]
-);
+use std::{cmp, ops};
 
-/// Wrapper for the third act message
-pub struct ActThree(
-	pub(super) [u8; ACT_THREE_LENGTH]
-);
+pub const ACT_ONE_TWO_LENGTH: usize = 50;
+pub const ACT_THREE_LENGTH: usize = 66;
+pub const EMPTY_ACT_ONE: ActOne = [0; ACT_ONE_TWO_LENGTH];
+pub const EMPTY_ACT_TWO: ActTwo = EMPTY_ACT_ONE;
+pub const EMPTY_ACT_THREE: ActThree = [0; ACT_THREE_LENGTH];
+type ActOne = [u8; ACT_ONE_TWO_LENGTH];
+type ActTwo = ActOne;
+type ActThree = [u8; ACT_THREE_LENGTH];
 
 /// Wrapper for any act message
-pub enum Act {
+pub(super) enum Act {
 	One(ActOne),
 	Two(ActTwo),
-	Three(ActThree),
+	Three(ActThree)
 }
 
 impl Act {
-	/// Convert any act into a byte vector
-	pub fn serialize(&self) -> Vec<u8> {
+	/// Returns the size of the underlying array
+	fn len(&self) -> usize {
+		self.as_ref().len()
+	}
+}
+
+impl From<ActBuilder> for Act {
+	/// Convert a finished ActBuilder into an Act
+	fn from(act_builder: ActBuilder) -> Self {
+		assert!(act_builder.is_finished());
+		act_builder.partial_act
+	}
+}
+
+impl ops::Deref for Act {
+	type Target = [u8];
+
+	/// Allows automatic coercion to slices in function calls
+	/// &Act -> &[u8]
+	fn deref(&self) -> &Self::Target {
 		match self {
 			&Act::One(ref act) => {
-				act.0.to_vec()
+				act
 			}
 			&Act::Two(ref act) => {
-				act.0.to_vec()
+				act
 			}
 			&Act::Three(ref act) => {
-				act.0.to_vec()
+				act
 			}
 		}
 	}
-}
\ No newline at end of file
+}
+
+impl AsRef<[u8]> for Act {
+	/// Allow convenient exposure of the underlying array through as_ref()
+	/// Act.as_ref() -> &[u8]
+	fn as_ref(&self) -> &[u8] {
+		&self
+	}
+}
+
+// Simple fill implementation for both almost-identical structs to deduplicate code
+// $act: Act[One|Two|Three], $input: &[u8]; returns &[u8] of remaining input that was not processed
+macro_rules! fill_impl {
+	($act:expr, $write_pos:expr, $input:expr) => {{
+		let fill_amount = cmp::min($act.len() - $write_pos, $input.len());
+
+		$act[$write_pos..$write_pos + fill_amount].copy_from_slice(&$input[..fill_amount]);
+
+		$write_pos += fill_amount;
+		&$input[fill_amount..]
+	}}
+}
+
+/// Light wrapper around an Act that allows multiple fill() calls before finally
+/// converting to an Act via Act::from(act_builder). Handles all of the bookkeeping
+/// and edge cases of the array fill
+pub(super) struct ActBuilder {
+	partial_act: Act,
+	write_pos: usize
+}
+
+impl ActBuilder {
+	/// Returns a new ActBuilder for Act::One
+	pub(super) fn new(empty_act: Act) -> Self {
+		Self {
+			partial_act: empty_act,
+			write_pos: 0
+		}
+	}
+
+	/// Fills the Act with bytes from input and returns the unprocessed bytes
+	pub(super) fn fill<'a>(&mut self, input: &'a [u8]) -> &'a [u8] {
```

Comment:
I think a somewhat more flexible approach than returning the unprocessed bytes would be returning the number of processed bytes, such that a caller would be able to call the function again from a new index, and know where to discard/drain data. It slightly shifts the responsibility of memory management, but avoids the awkwardness of lifetime annotations and having multiple borrowed references to the same data.